copy-paste of the async-race project for the epamJSxStackLab (#1589)
rolling-scopes-school/tasks: epam/async-race.md

# Welcome to the test task!
Below is a description of the test task. Please read it carefully. The ability to follow the requirements is important to us.
You should complete the task before proceeding to the recruiter's interview and subsequent technical interview.
During the technical interview, the interviewer may discuss your solution. 

## Task "Async Race"
#### Try to imagine the scenario. 

> Your customer has an idea to spend his money. He has hired some engineers who installed radio-controlled equipment :radio: in real cars :car:, to be precise, all of his collection of cars. He is eager to create a kind of drag-racing competition in order to discover which car is the fastest.

Reviewer: i would make it gender-neutral

What are you waiting for? Let's start coding!

## 🎯 Objectives

- **Create a Single Page Application (SPA)**: Manage a collection of cars, operate their engines, and showcase race statistics in an engaging, interactive SPA.
- **Revive the Project**: Drawing inspiration from a demo video and a server mock, reconstruct the UI and breathe life back into this ambitious project.
- **Outpace the Competition**: Implement the project with speed and efficiency, ensuring we launch before our rivals, showcasing our innovative solution to car management and racing competitions.

## 🚗 Functional Requirements

### Basic Structure

- Two main views: "Garage" and "Winners", each with their name, page number, and a count of items in the database.
- Persistent view state between switches, maintaining user input and pagination.

### Garage View

> - CRUD operations for cars with "name" and "color" attributes.

Reviewer: ```suggestion
- (CRUD)[https://en.wikipedia.org/wiki/Create,_read,_update_and_delete] operations for cars with "name" and "color" attributes.
```

- Color selection from an RGB palette with a preview of the car in the chosen color.
- Pagination to display cars (7 per page) and a feature to generate 100 random cars at once.

### Car Animation

- Start/stop engine buttons with corresponding animations and handling of engine states.
- Adaptive animations that work on screens as small as 500px.

### Race Animation

- A button to start a race for all cars on the current page.
- A reset button to return all cars to their starting positions.
- Display the winner's name upon race completion.

### Winners View

- Display winning cars with their image, name, number of wins, and best time.
- Pagination and sorting capabilities by wins and best times.

## 📜 Non-functional Requirements

> - **Framework Usage**:  Development must utilize either React or Angular.

Reviewer: let's define minimal versions just in case. E.g. React 18 or Angular 17

> - **TypeScript Mandatory**: Strong typing throughout, with no use of "any".

Reviewer: Also probably makes sense to ask have the following settings in tsconfig.json to make sure strong typing enforced.

"strict": true,
"noImplicitAny": true,

- **Modular Architecture**: Clear separation of concerns between API interaction, UI rendering, and state management.
- **SPA Implementation**: All content must be generated using TypeScript.

> - **Code Quality**: Adherence to Airbnb's ESLint configuration, with functions limited to 40 lines and avoidance of magic numbers/strings.

Reviewer: let's add link to a package to avoid people to pick something different: https://www.npmjs.com/package/eslint-config-airbnb

- These criteria emphasize the importance of clean code, proper architecture, and adherence to modern development practices. Intreviewers will use these guidelines to assess the technical execution and overall quality of the project.

## 🛠️ Technical Implementation

- Implement CRUD operations for cars using the provided server mock.
- Design UI elements for car management and race controls.
- Utilize fetch for server communication, and handle promises for asynchronous tasks.
- Create animations for car movements using TypeScript and CSS.
- Ensure responsiveness and compatibility across different devices and browsers.

## Requirements to commits and repo
### Commit Guidelines

> - Follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0-beta.2/) format.

Reviewer: ```suggestion
- Follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) format.
```

- Use lowercase for commit types (`init`, `feat`, `fix`, `refactor`, `docs`, etc.).
- Write in present tense (e.g., "add feature") and imperative mood (e.g., "move cursor to...").

#### Commit Examples
- `init:` Start a project/task (e.g., `init: start youtube-task`).
- `feat:` Introduce new features (e.g., `feat: add zoom support`).
- `fix:` Correct bugs in existing features (e.g., `fix: correct data loading`).
- `refactor:` Improve the code without altering functionality (e.g., `refactor: rename vars`).
- `docs:` Update documentation (e.g., `docs: update readme`).

### README.md 
When your solution is implemented, please include the checklist below in the `README.md`. Check all implemented features, then calculate your score and put it at the top of the `README.md`.

> Also, please deploy the UI to gh-pages, AWS Amplify, Netlify, or a similar service, and put the link at the top of the `README.md` (alone with the calculated score).

Reviewer: let's remove Amplify. It's bad for juniors.

gh-pages, Vercel, Netlify, Cloudflare pages

## Checklist

## 🚀 UI Deployment

> - [ ] **Deployment Platform:** Successfully deploy the UI on one of the following platforms: GitHub Pages, AWS Amplify, Netlify, or a similar service.

Reviewer: ```suggestion
- [ ] **Deployment Platform:** Successfully deploy the UI on one of the following platforms: GitHub Pages, Netlify, Vercel, Cloudflare Pages or a similar service.
```
